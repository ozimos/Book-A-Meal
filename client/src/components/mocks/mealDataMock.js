export const rows = [
  { id: 'abc', price: 1200 }, { id: 'def', price: 2000 }
];

export const allMeals = {
  data: { rows }

};

export const meal = {
  data: { id: 'mealId',
    price: 200,
    description: 'nice',
    imageUrl: 'nice',
    title: 'nice'
  }
};
