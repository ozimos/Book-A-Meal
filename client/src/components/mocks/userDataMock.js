export const userData = {
  data: { id: '123', email: 'asdf' },
  message: 'abc',
  token: 'abc'
};
export const userInput = {
  data: { email: 'asdf' }
};
export const userIncludes = {
  email: 'asdf',
  firstName: 'first',
  lastName: 'last'
};
