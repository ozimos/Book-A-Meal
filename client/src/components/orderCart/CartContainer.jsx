import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { orderActions } from '../../redux/actions';


import '../../../public/styles/book_a_meal.css';
import '../../../public/styles/cart_layout.scss';


class CartContainer extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      orderQuantity: {},
      prevPropsOrderId: this.props.orderId
    };
  }
  static getDerivedStateFromProps(nextProps, prevState) {
    const newOrderQuantity = {};
    if (nextProps.orderId !== prevState.prevPropsOrderId) {
      nextProps.order
        .forEach((elem) => { newOrderQuantity[elem.id] = elem.MealOrders.quantity; });
      return {
        prevPropsOrderId: nextProps.orderId,
        orderQuantity: newOrderQuantity };
    }
    return null;
  }
  componentDidMount() {
    if (this.props.orderId) {
      const newOrderQuantity = {};

      this.props.order
        .forEach((elem) => { newOrderQuantity[elem.id] = elem.MealOrders.quantity; });
      // eslint-disable-next-line
      this.setState({ orderQuantity: newOrderQuantity });

    }
  }
  setQuantity = (e, id) => {
    const inputValue = e.target.value;
    this.setState(prevState =>
      ({ orderQuantity: { ...prevState.orderQuantity, [id]: inputValue } }));
  };
 placeOrder = async () => {
   const actualOrder = this.props.order
     .map(meal => ({ id: meal.id,
       quantity: this.state.orderQuantity[meal.id] || 1 }));
   if (this.props.orderId) {
     await this.props
       .dispatch(orderActions.updateOrder({ meals: actualOrder }, this.props.orderId));
   } else {
     await this.props.dispatch(orderActions.postOrder({ meals: actualOrder }));
   }
   /* eslint no-alert: off */
   if (!this.props.orderError) {
     alert('Order has been created');
     this.props.clearCart();
   }
   if (this.props.orderError) {
     alert(this.props.orderError);
   }
 }

 render() {
   const { MealRow, order, ...rest } = this.props;
   const reducer = (total, meal) => total + (meal.price * (this.state.orderQuantity[meal.id] || 1));
   const calcTotal = () => this.props.order
     .reduce(reducer, 0);
   const total = calcTotal();
   return (
     <div className={rest.addClass ? `${rest.addClass}` : ''}>
       <h3>Order Cart</h3>
       <div className="table-responsive">
         <table className="table">
           <thead>
             <tr>
               <th>Remove Item</th>
               <th>Meal Title</th>
               <th>Price (&#x20a6;)</th>
               <th>Quantity</th>
               <th>Subtotal (&#x20a6;)</th>
             </tr>
           </thead>
           <tbody>
             {order.map(meal => (<MealRow
               key={meal.id}
               {...meal}
               setQuantity={this.setQuantity}
               quantity={(this.state.orderQuantity[meal.id] || '')}
               {...rest}
             />))
        }
           </tbody>
           <caption>
             <div className="row">
               <div className="col-3 mr-auto">&#9776; Your Order</div>
               <div className="col-2"> Total </div>
               <div className="col-2"> {total}</div>
               <div className="col2">
                 <button className="btn btn-cart" onClick={rest.clearCart}>
                   <p>Clear Cart</p>
                 </button>
               </div>
               <div className="col2">
                 <button className="btn btn-cart" onClick={this.placeOrder}>
                   {this.props.orderId ? <p>Modify Order</p> : <p>Place New Order</p>}
                 </button>
               </div>
             </div>
           </caption>
         </table>
       </div>
     </div>
   );
 }
}

CartContainer.defaultProps = {
  addClass: '',
  orderError: '',
  orderId: ''
};
CartContainer.propTypes = {
  order: PropTypes.arrayOf(PropTypes.object).isRequired,
  MealRow: PropTypes.func.isRequired,
  addClass: PropTypes.string,
  orderId: PropTypes.string,
  orderError: PropTypes.string,
  clearCart: PropTypes.func.isRequired,
  dispatch: PropTypes.func.isRequired,
};
const mapStateToProps = state => ({
  orderError: state.orderReducer.orderError,
  connecting: state.orderReducer.connecting,
});
export { CartContainer };
export default connect(mapStateToProps)(CartContainer);

