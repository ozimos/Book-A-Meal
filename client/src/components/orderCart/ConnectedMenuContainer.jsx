import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { menuActions } from '../../redux/actions';

import '../../../public/styles/cart_layout.scss';


class MenuContainer extends React.Component {
 placeOrder = async () => {

   if (this.props.postMenu) {
     await this.props.postMenu(this.props.menu);
   } else {
     await this.props.dispatch(menuActions.postMenu(this.props.menu));
   }

   if (!this.props.menuError) {
     const message = this.props.postMenu ? 'added to' : 'removed from';
     this.props.notify(`The selected meals have been ${message} the menu`);
     this.props.clearMenu();
   }

   if (this.props.menuError) {
     this.props.notify(this.props.menuError);
   }
 }

 render() {
   const { menu, ...rest } = this.props;

   return (
     <div className={rest.addClass ? `${rest.addClass}` : ''}>
       <div className="flexbox">
         <h5>Menu Editor</h5>
         <button className="btn title-button" onClick={rest.closeMenuModal}>
                &#10006;
         </button>
       </div>

       <div className="scroll">
         {menu.map(meal =>
             (
               <div key={meal.id} className="flexbox">
                 <span
                   className="text_left long_string"
                   key={meal.id}
                 >
                   {meal.title}
                 </span>
                 <button
                   className="btn title-button"
                   onClick={() => rest.removeFromMenu(meal.id)}
                 >
               Remove
                 </button>
               </div>
               ))
        }
       </div>

       <div className="flexbox">

         <button className="btn btn-cart" onClick={this.placeOrder}>
              Post meals to Menu
         </button>

         <button className="btn btn-cart" onClick={rest.clearMenu}>
                   Clear Menu
         </button>

       </div>

     </div>
   );
 }
}

MenuContainer.defaultProps = {
  addClass: '',
  menuError: '',
};

MenuContainer.propTypes = {
  menu: PropTypes.arrayOf(PropTypes.object).isRequired,
  addClass: PropTypes.string,
  menuError: PropTypes.string,
  clearMenu: PropTypes.func.isRequired,
  notify: PropTypes.func.isRequired,
  dispatch: PropTypes.func.isRequired,
  postMenu: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  menuError: state.menuReducer.menuError,
  connecting: state.menuReducer.connecting,
});

export { MenuContainer };
export default connect(mapStateToProps)(MenuContainer);

