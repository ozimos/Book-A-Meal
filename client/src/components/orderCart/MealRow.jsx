import React from 'react';
import PropTypes from 'prop-types';
import { hot } from 'react-hot-loader';

const MealRow = ({ id, title, price, quantity, setQuantity, removeFromCart }) => {
  const subTotal = price * (quantity || 1);
  return (
    <React.Fragment>
      <tr>
        <td>
          <button className="btn title-button icon" onClick={() => removeFromCart(id)}>
                &#10006;
          </button>
        </td>
        <td>{title}</td>
        <td>{price}</td>
        <td>
          <input
            className="formControl"
            type="number"
            placeholder="1"
            onChange={event => setQuantity(event, id)}
            value={quantity}
            style={{ width: '40px' }}
          />
        </td>
        <td>{subTotal}</td>
      </tr>
    </React.Fragment>
  );
};

MealRow.propTypes = {
  id: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  price: PropTypes.number.isRequired,
  quantity: PropTypes.oneOfType(PropTypes.string, PropTypes.number).isRequired,
  removeFromCart: PropTypes.func.isRequired,
  setQuantity: PropTypes.func.isRequired,
};
export default hot(module)(MealRow);
