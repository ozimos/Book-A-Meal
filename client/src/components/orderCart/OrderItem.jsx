import React from 'react';
import PropTypes from 'prop-types';
import format from 'date-fns/format';
import differenceInMinutes from 'date-fns/difference_in_minutes';
import isToday from 'date-fns/is_today';
import '../../../public/styles/order_item.scss';

const OrderItem = ({ id, Meals, addOrder, position, updatedAt }) => {
  const details = Meals.map(meal => (
    <div key={meal.id} className="row">
      <div className="col s7 text_left long_string left_pad">{meal.title}</div>
      <div className="col s3 text_left left_pad">{meal.price}</div>
      <div className="col s2 text_left">{meal.MealOrders.quantity}</div>
    </div>));
  const hasEditMinutes = (parseInt(process.env.ORDER_EDIT_MINUTES, 10) || 15)
       - differenceInMinutes(new Date(), updatedAt) > 0;

  return (
    <div className="row order_item" style={{ borderBottom: '1px solid' }}>
      <div className="col s1 text_left">{position}</div>
      <div className="col s9">
        {details}
      </div>
      <div className="col s1 text_left">{format(updatedAt, 'h:m A')}</div>
      <div className="col s1 text_right">
        {
          isToday(updatedAt) && hasEditMinutes &&
          <button onClick={() => addOrder(id)} className="btn title-button">Edit</button>
        }
      </div>
    </div>
  );
};

OrderItem.propTypes = {
  id: PropTypes.string.isRequired,
  updatedAt: PropTypes.string.isRequired,
  position: PropTypes.number.isRequired,
  Meals: PropTypes.arrayOf(PropTypes.object).isRequired,
  addOrder: PropTypes.func.isRequired
};
export default OrderItem;

