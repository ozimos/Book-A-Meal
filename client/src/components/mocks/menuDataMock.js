import { allMeals } from './mealDataMock';

export default {
  data: { Meals: allMeals.data }
};