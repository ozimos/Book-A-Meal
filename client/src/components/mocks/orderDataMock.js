import { allMeals } from './mealDataMock';

const Meals = allMeals.data.map(meal => ({ ...meal, MealOrders: { quantity: 2 } }));
export const order = {
  data: { id: 'orderId',
    userId: 'userId',
    Meals,
    mealList: ['abc', 'def'],
    quantityList: [2, 2] }
};
export const allOrders = {
  data: [{ ...order.data, id: 'order1' }, { ...order.data, id: 'order2' }]
};