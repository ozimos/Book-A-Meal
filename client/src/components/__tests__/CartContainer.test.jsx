/* global React:false, shallow:false toJson:false */

import { CartContainer } from '../orderCart/CartContainer';
import { allOrders } from '../mocks/orderDataMock';

describe('CartContainer Component', () => {

  it('should render without throwing an error', () => {
    expect(shallow(<CartContainer
      order={allOrders.data}
      MealRow={jest.fn()}
      clearCart={jest.fn()}
      dispatch={jest.fn()}
      orderId=""
    />)
      .exists(<table className="table" />)).toBe(true);
  });
  it('renders correctly', () => {
    const wrapper = shallow(<CartContainer
      order={allOrders.data}
      MealRow={jest.fn()}
      clearCart={jest.fn()}
      dispatch={jest.fn()}
      orderId=""
    />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

