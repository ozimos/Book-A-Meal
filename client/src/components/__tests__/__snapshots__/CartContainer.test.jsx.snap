// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartContainer Component renders correctly 1`] = `
<div
  className=""
>
  <h3>
    Order Cart
  </h3>
  <div
    className="responsive-table"
  >
    <table
      className="table"
    >
      <thead>
        <tr>
          <th>
            Remove Item
          </th>
          <th>
            Meal Title
          </th>
          <th>
            Price (₦)
          </th>
          <th>
            Quantity
          </th>
          <th>
            Subtotal (₦)
          </th>
        </tr>
      </thead>
      <tbody>
        <mockConstructor
          Meals={
            Array [
              Object {
                "MealOrders": Object {
                  "quantity": 2,
                },
                "id": "abc",
                "price": 1200,
              },
              Object {
                "MealOrders": Object {
                  "quantity": 2,
                },
                "id": "def",
                "price": 2000,
              },
            ]
          }
          User={
            Object {
              "email": "asdf",
              "firstName": "first",
              "lastName": "last",
            }
          }
          addClass=""
          clearCart={[MockFunction]}
          dispatch={[MockFunction]}
          id="order1"
          key="order1"
          mealList={
            Array [
              "abc",
              "def",
            ]
          }
          orderError=""
          orderId=""
          quantity=""
          quantityList={
            Array [
              2,
              2,
            ]
          }
          setQuantity={[Function]}
          userId="userId"
        />
        <mockConstructor
          Meals={
            Array [
              Object {
                "MealOrders": Object {
                  "quantity": 2,
                },
                "id": "abc",
                "price": 1200,
              },
              Object {
                "MealOrders": Object {
                  "quantity": 2,
                },
                "id": "def",
                "price": 2000,
              },
            ]
          }
          User={
            Object {
              "email": "asdf",
              "firstName": "first",
              "lastName": "last",
            }
          }
          addClass=""
          clearCart={[MockFunction]}
          dispatch={[MockFunction]}
          id="order2"
          key="order2"
          mealList={
            Array [
              "abc",
              "def",
            ]
          }
          orderError=""
          orderId=""
          quantity=""
          quantityList={
            Array [
              2,
              2,
            ]
          }
          setQuantity={[Function]}
          userId="userId"
        />
      </tbody>
      <caption>
        <div
          className="row"
        >
          <div
            className="col s3 mr-auto"
          >
            ☰ Your Order
          </div>
          <div
            className="col s2"
          >
             Total 
          </div>
          <div
            className="col s2"
          >
             
            NaN
          </div>
          <div
            className="col s2"
          >
            <button
              className="btn btn-cart"
              onClick={[MockFunction]}
            >
              <p>
                Clear Cart
              </p>
            </button>
          </div>
          <div
            className="col s2"
          >
            <button
              className="btn btn-cart"
              onClick={[Function]}
            >
              <p>
                Place New Order
              </p>
            </button>
          </div>
        </div>
      </caption>
    </table>
  </div>
</div>
`;
